server/list.go

```diff
@@ -0,0 +1,239 @@
+	}
+
+	return nil
+}
+
+func (l *listManager) Send(senderID string, receiverID string, message string) (string, error) {
+	senderItem := newItem(message)
+	l.store.AddItem(senderItem)
+	receiverItem := newItem(message)
+	l.store.AddItem(receiverItem)
+
+	appErr := l.store.Add(senderID, senderItem.ID, OutListKey, receiverID, receiverItem.ID)
+	if appErr != nil {
+		return "", appErr
+	}
+
+	appErr = l.store.Add(receiverID, receiverItem.ID, InListKey, senderID, senderItem.ID)
+	if appErr != nil {
+		return "", appErr
+	}
+
+	return receiverItem.ID, nil
+}
+
+func (l *listManager) Get(userID string, listID string) ([]*ExtendedItem, error) {
+	oes, err := l.store.GetList(userID, listID)
+	if err != nil {
+		return nil, err
+	}
+
+	extendedItems := []*ExtendedItem{}
+	for _, oe := range oes {
+		item, err := l.store.GetItem(oe.ItemID)
+		if err != nil {
+			continue
+		}
+
+		extendedItem := l.extendItemInfo(item, oe)
+		extendedItems = append(extendedItems, extendedItem)
+	}
+
+	return extendedItems, nil
+}
+
+func (l *listManager) Complete(userID string, itemID string) (todoMessage string, foreignUserID string, outErr error) {
+	itemList, oe, _ := l.store.GetItemListAndOrder(userID, itemID)
+	if oe == nil {
+		return "", "", fmt.Errorf("cannot find element")
+	}
+
+	if err := l.store.Remove(userID, itemID, itemList); err != nil {
+		return "", "", err
+	}
+
+	l.store.RemoveItem(itemID)
+
+	if oe.ForeignUserID == "" {
+		return "", "", nil
+	}
+
+	l.store.Remove(oe.ForeignUserID, oe.ForeignItemID, OutListKey)
+	item, err := l.store.GetItem(oe.ForeignItemID)
+	if err != nil {
+		return "", "", nil
+	}
+
+	l.store.RemoveItem(oe.ForeignItemID)
+
+	return item.Message, oe.ForeignUserID, nil
+}
+
+func (l *listManager) Enqueue(userID string, itemID string) (todoMessage string, foreignUserID string, outErr error) {
+	item, err := l.store.GetItem(itemID)
+	if err != nil {
+		return "", "", err
+	}
+
+	oe, _, err := l.store.GetItemOrder(userID, itemID, InListKey)
+	if err != nil {
+		return "", "", err
+	}
+	if oe == nil {
+		return "", "", fmt.Errorf("element order not found")
+	}
+
+	err = l.store.Add(userID, itemID, MyListKey, oe.ForeignUserID, oe.ForeignItemID)
+	if err != nil {
+		return "", "", err
+	}
+
+	err = l.store.Remove(userID, itemID, InListKey)
+	if err != nil {
+		l.store.Remove(userID, itemID, MyListKey)
+		return "", "", err
+	}
+
+	return item.Message, oe.ForeignUserID, nil
+}
+
+func (l *listManager) Remove(userID string, itemID string) (todoMessage string, foreignUserID string, isSender bool, outErr error) {
+	itemList, oe, _ := l.store.GetItemListAndOrder(userID, itemID)
+	if oe == nil {
+		return "", "", false, fmt.Errorf("cannot find element")
+	}
+
+	if err := l.store.Remove(userID, itemID, itemList); err != nil {
+		return "", "", false, err
+	}
+
+	l.store.RemoveItem(itemID)
+
+	if oe.ForeignUserID == "" {
+		return "", "", false, nil
+	}
+
+	list, _, _ := l.store.GetItemListAndOrder(oe.ForeignUserID, oe.ForeignItemID)
```

Comment:
I wonder -- can you think of a way to change how the items, list and order are stored? There is probably a simpler way... What do you think? (It might be too much work though, so feel free to say its out of scope.)